abe/abe.go

```diff
@@ -0,0 +1,472 @@
+package abe
+
+import (
+	"fmt"
+	"math/big"
+
+	"strconv"
+	"strings"
+
+	"github.com/cloudflare/bn256"
+	"github.com/fentec-project/gofe/data"
+	"github.com/fentec-project/gofe/sample"
+)
+
+// abeParams represents configuration parameters for the ABE scheme instance.
+type abeParams struct {
+	l int      // number of attributes
+	p *big.Int // order of the elliptic curve
+}
+
+// Abe represents an ABE scheme.
+type Abe struct {
+	Params *abeParams
+}
+
+// NewAbe configures a new instance of the scheme.
+// It accepts l the number of attributes possibly used in
+// the scheme. Attributes' names will be considered as
+// elements of a set {0, 1,..., l-1}.
+func newAbe(l int) *Abe {
+	return &Abe{Params: &abeParams{
+		l: l, // number of attributes in the whole universe
+		p: bn256.Order,
+	}}
+}
+
+// AbePubKey represents a public key of an ABE scheme.
+type AbePubKey struct {
+	t data.VectorG2
+	y *bn256.GT
+}
+
+// GenerateMasterKeys generates a new set of public keys, needed
+// for encrypting data, and secret keys needed for generating keys
+// for decryption.
+func (a Abe) GenerateMasterKeys() (*AbePubKey, data.Vector, error) {
+	// todo: should we put public key into Abe struct?
+	sampler := sample.NewUniform(a.Params.p)
+	sk, err := data.NewRandomVector(a.Params.l+1, sampler)
+	if err != nil {
+		return nil, nil, err
+	}
+	t := sk[:a.Params.l].MulG2()
+	y := new(bn256.GT).ScalarBaseMult(sk[a.Params.l])
+
+	return &AbePubKey{t: t, y: y}, sk, nil
+}
+
+// AbeCipher represents a ciphertext of an ABE scheme.
+type AbeCipher struct {
+	gamma     []int         // the set of attributes that can be used for policy of decryption
+	attribToI map[int]int   // a map that connects the attributes in gamma with elements of e
+	e0        *bn256.GT     // the first part of the encryption
+	e         data.VectorG2 // the second part of the encryption
```

Comment:
`e0` and `e` were chosen as names because of the naming in the relevant paper(s)? They are a bit nondescriptive.

Reply:
Yes, it is as in the paper and is hard to describe them, encryption is two values, one an element of a GT and second a vector of G2 elements.